src/pdu_loop/frame_element.rs

```diff
@@ -0,0 +1,490 @@
+use crate::{
+    command::Command,
+    error::{Error, PduError},
+    pdu_loop::{frame_header::FrameHeader, pdu_flags::PduFlags},
+    ETHERCAT_ETHERTYPE, MASTER_ADDR,
+};
+use cookie_factory::{
+    bytes::{le_u16, le_u8},
+    combinator::{skip, slice},
+    gen_simple, GenError,
+};
+use core::{
+    future::Future,
+    marker::PhantomData,
+    mem,
+    ops::Deref,
+    ptr::{addr_of, addr_of_mut, NonNull},
+    sync::atomic::Ordering,
+    task::{Poll, Waker},
+};
+use packed_struct::PackedStruct;
+use smoltcp::wire::{EthernetAddress, EthernetFrame};
+
+use super::PduResponse;
+
+#[atomic_enum::atomic_enum]
+#[derive(PartialEq, Default)]
+pub enum FrameState {
+    // SAFETY: Because we create a bunch of `Frame`s with `MaybeUninit::zeroed`, the `None` state
+    // MUST be equal to zero. All other fields in `Frame` are overridden in `replace`, so there
+    // should be no UB there.
+    /// The frame is ready to be claimed
+    #[default]
+    None = 0,
```

Comment:
Note: only create_frame is allowed to touch "empty frames", this is a "free allocation block".